cgi/suggest.pl

```diff
@@ -42,84 +42,146 @@
 }
 
 my $original_lc = $search_lc;
 
+# if search begins with a language code, use it for search
 if ($term =~ /^(\w\w):/) {
 	$search_lc = $1;
 	$term = $';
 }
 
+# max results
+my $limit = 25;
+# superseed by request parameter
+if (defined param('limit')) {
+	# we put a hard limit however
+	$limit = min(int(param('limit')), 400);
+}
+
+
 my @suggestions = (); # Suggestions starting with the term
 my @suggestions_c = (); # Suggestions containing the term
+my @suggestions_f = (); # fuzzy suggestions
 
 my $cache_max_age = 0;
-my $limit = 25;
-my $i = 0;
+my $suggestions_count = 0;
+
+
+# search for emb codes
 if ($tagtype eq 'emb_codes') {
 	my $stringid = get_string_id_for_lang("no_language", normalize_packager_codes($term));
 	my @tags = sort keys %packager_codes;
 	foreach my $canon_tagid (@tags) {
 		next if $canon_tagid !~ /^$stringid/;
 		push @suggestions, normalize_packager_codes($canon_tagid);
-		last if ++$i >= $limit;
+		last if ++$suggestions_count >= $limit;
 	}
+	# add cache to request
 	$cache_max_age = 3600;
 }
 else {
-	my $stringid = get_string_id_for_lang($search_lc, $string) . get_string_id_for_lang($search_lc, $term);
+	# search for term in a taxonomy
+
+	# normalize string and term
+	my $stringid = get_string_id_for_lang($search_lc, $string) . "-" . get_string_id_for_lang($search_lc, $term);
+	# remove eventual leading or ending "-"
+	$stringid =~ s/^-//;
+	$stringid =~ s/^-$//;
+	# fuzzy match whole words with eventual inter-words
+	my $fuzzystringid = join(".*", split("-", $stringid));
+	# all tags can be retrieve from the $translations_to hash
 	my @tags = sort keys %{$translations_to{$tagtype}} ;
 	foreach my $canon_tagid (@tags) {
-		
+		# just_synonyms are not real entries
 		next if defined $just_synonyms{$tagtype}{$canon_tagid};
-		
-		my $tag;
-		my $tagid;
-		
+
+		my $tag;  # this is the content string
+		my $tagid;  # this is the tag
+
+		# search if the tag exists in target language
 		if (defined $translations_to{$tagtype}{$canon_tagid}{$search_lc}) {
-		
+
 			$tag = $translations_to{$tagtype}{$canon_tagid}{$search_lc};
+			# TODO: explain why $tagid can be different from $canon_tagid
 			$tagid = get_string_id_for_lang($search_lc, $tag);
-			
+
+			# add language prefix if we are not searching current interface language
 			if (not ($search_lc eq $original_lc)) {
 				$tag = $search_lc . ":" . $tag;
 			}
 		}
+		# also search for special language code "xx" which is universal
 		elsif (defined $translations_to{$tagtype}{$canon_tagid}{xx}) {
 			$tag = $translations_to{$tagtype}{$canon_tagid}{xx};
 			$tagid = get_string_id_for_lang("xx", $tag);
 		}
-		
+
 		if (defined $tag) {
-		
-			next if $tagid !~ /$stringid/;
-			
-			if ($tag =~ /^$stringid/i) {
+			# matching at start
+			if ($tagid =~ /^$stringid/) {
 				push @suggestions, $tag;
+				$suggestions_count++;
 			}
-			else {
+			# matching inside
+			elsif ($tagid =~ /$stringid/) {
 				push @suggestions_c, $tag;
+				$suggestions_count++;
```

Comment:
One issue with doing that (which was already present before) is that we will hit the limit with entries that don't start with the searched string, even though we may have some.

e.g. just searching with one letter:

![image](https://user-images.githubusercontent.com/8158668/173316037-92c9a59b-2a84-4a50-90b4-6d3810a236a4.png)

So we could create a separate counter for suggestions matching inside instead.